### ISSUE TITLE
inotify: leaking inotify_kevent_t causes an IN_Q_OVERFLOW event very soon

### ISSUE THREAD

**Issue by youzhongyang** [joyent/illumos-joyent#131]
Recently we tested Samba 4.4.6 on latest SmartOS and found that after some time, 'change notification' stopped working.

Further investigation revealed the cause of the problem: when the Samba 'notifyd' smbd process frequently calls inotify_add_watch() and inotify_rm_watch() for hundreds of folders, very soon events similar to the following one were piled up and exceeded 'inotify_maxevents':

<pre>
inotify_kevent_t {
    inotify_kevent_t *ine_next = 0xffffd0c57524f800
    struct inotify_event ine_event = {
        int32_t wd = 0x2
        uint32_t mask = 0x8000
        uint32_t cookie = 0
        uint32_t len = 0
        char [0] name = [ ... ]
    }
}
</pre>

This issue can be reproduced using the following test program:

``` c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/inotify.h>
#ifdef __sun__
#include <sys/filio.h>
#endif

#include <unistd.h>
#include <stropts.h>
#include <poll.h>
#include <dirent.h>
#include <sys/stat.h>

#define EVENT_SIZE  ( sizeof (struct inotify_event) )
#define EVENT_BUF_LEN     ( 1024 * ( EVENT_SIZE + 64 ) )

void print_inotify_event(struct inotify_event *event)
{
    if ( event->len ) {
        printf("   :: mask %d\n", event->mask);
        if ( event->mask & IN_CREATE ) {
            if ( event->mask & IN_ISDIR ) {
                printf( "   -> New directory %s created\n", event->name );
            }
            else {
                printf( "   -> New file %s created\n", event->name );
            }
        }
        else if ( event->mask & IN_DELETE ) {
            if ( event->mask & IN_ISDIR ) {
                printf( "   -> Directory %s deleted\n", event->name );
            }
            else {
                printf( "   -> File %s deleted\n", event->name );
            }
        }
        else if ( event->mask & IN_MOVED_FROM ) {
            printf( "   -> %s MOVED_FROM\n", event->name );
        }
        else if ( event->mask & IN_MOVED_TO ) {
            printf( "   -> %s MOVED_TO\n", event->name );
        }
        else if ( event->mask & IN_ATTRIB ) {
            printf( "   -> %s ATTRIB\n", event->name );
        }     
        else if ( event->mask & IN_MODIFY ) {
            printf( "   -> %s MODIFY\n", event->name );
        }        
        else {
            printf( "   -> MASK %d -> %s\n", event->mask, event->name);
        }
    }
}

void check_inotify_data(int fd)
{
    int length, i = 0;
    char buffer[EVENT_BUF_LEN];
    int bufsize;
    int ret;
    struct inotify_event *event;  
    struct pollfd fds[1];

    fds[0].fd = fd;
    fds[0].events = POLLIN;

    /* check if there's data available for reading */
    ret = poll(fds, 1, 1000);
    if(ret <= 0) {
        return;
    }

    bufsize = 0;
    /* check how many bytes available for reading */
    ret = ioctl(fd, FIONREAD, &bufsize); 

    /* Read the data. Normally the program can allocate a buffer based
       on the size returned by above iotcl call. Here we just use a static
       buffer 
    */
    length = read( fd, buffer, EVENT_BUF_LEN ); 

    /* checking for error */
    if ( length < 0 ) {
        perror( "read" );
        return;
    }  
    printf("got %d bytes\n", length);

    /*
        Actually read the list of change events. 
        Here, read the change event one by one and process it accordingly.
    */
    i = 0;
    while ( i < length ) {     
        event = ( struct inotify_event * ) &buffer[ i ];
        print_inotify_event(event);
        i += EVENT_SIZE + event->len;
    }

    return;
}

int add_inotify_watch(int fd, char *path)
{
    int mask = IN_MASK_ADD | IN_ONLYDIR | IN_CREATE | IN_DELETE | IN_MOVED_FROM | IN_MOVED_TO | IN_MODIFY | IN_ATTRIB;
    int wd;

    wd = inotify_add_watch(fd, path, mask);
    if (wd == -1) perror("inotify_add_watch");
    sleep(1);
    check_inotify_data(fd);
    return wd;
}

int remove_inotify_watch(int fd, int wd)
{
    int rc = inotify_rm_watch(fd, wd);
    if (rc) perror("inotify_rm_watch");
    sleep(1);
    check_inotify_data(fd); 
    return rc;
}

int make_dir(char *path)
{
    int rc;

    rc = mkdir(path, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
    if (rc) perror("mkdir");
    return rc;
}

int main(int argc, char *argv[] )
{
    int fd;
    char *folder_root = ".";
    char *folder_scratch = "./scratch";
    char *folder1 = "./scratch/folder1";
    char *folder2 = "./scratch/folder2";
    char *folder11 = "./scratch/folder1/folder11";
    char *folder21 = "./scratch/folder2/folder21";
    int wd_root = -1, wd_scratch = -1;
    int wd_folder1 = -1, wd_folder2 = -1;
    int wd_folder11 = -1, wd_folder21 = -1;

    /* create folders */
    if(make_dir(folder_scratch)) exit(1);
    if(make_dir(folder1)) exit(1);
    if(make_dir(folder2)) exit(1);
    if(make_dir(folder11)) exit(1);
    if(make_dir(folder21)) exit(1);

    /* create the INOTIFY instance */
    fd = inotify_init();
    if ( fd < 0 ) {
        perror( "inotify_init" );
        exit(1);
    }

    /* watch ./scratch */
    wd_scratch = add_inotify_watch(fd, folder_scratch);
    if (wd_scratch == -1) goto end;

    /* watch ./ */
    wd_root = add_inotify_watch(fd, folder_root);
    if (wd_root == -1) goto end;

    /* watch ./scratch/folder1 */
    wd_folder1 = add_inotify_watch(fd, folder1);
    if (wd_folder1 == -1) goto end;

    /* remove_watch ./ */
    if(remove_inotify_watch(fd, wd_root)) goto end;
    wd_root = -1;

    /* watch ./scratch/folder1/folder11 */
    wd_folder11 = add_inotify_watch(fd, folder11);
    if (wd_folder11 == -1) goto end;

    /* remove_watch ./scratch */
    if(remove_inotify_watch(fd, wd_scratch)) goto end;
    wd_scratch = -1;

    /* watch ./scratch */
    wd_scratch = add_inotify_watch(fd, folder_scratch);
    if (wd_scratch == -1) goto end;

    /* remove_watch ./scratch/folder1 */
    if(remove_inotify_watch(fd, wd_folder1)) goto end;
    wd_folder1 = -1;

    /* remove_watch ./scratch/folder1/folder11 */
    if(remove_inotify_watch(fd, wd_folder11)) goto end;
    wd_folder11 = -1;

    /* watch ./ */
    wd_root = add_inotify_watch(fd, folder_root);
    if (wd_root == -1) goto end;

    /* watch ./scratch/folder2 */
    wd_folder2 = add_inotify_watch(fd, folder2);
    if (wd_folder2 == -1) goto end;

    /* remove_watch ./ */
    if(remove_inotify_watch(fd, wd_root)) goto end;
    wd_root = -1;

    /* watch ./scratch/folder2/folder21 */
    wd_folder21 = add_inotify_watch(fd, folder21);
    if (wd_folder21 == -1) goto end;

    /* remove_watch ./scratch */
    if(remove_inotify_watch(fd, wd_scratch)) goto end;
    wd_scratch = -1;

    /* watch ./scratch */
    wd_scratch = add_inotify_watch(fd, folder_scratch);
    if (wd_scratch == -1) goto end;

    /* remove_watch ./scratch/folder2 */
    if(remove_inotify_watch(fd, wd_folder2)) goto end;
    wd_folder2 = -1;

    /* remove_watch ./scratch/folder2/folder21 */
    if(remove_inotify_watch(fd, wd_folder21)) goto end;
    wd_folder21 = -1;

    /* watch ./ */
    wd_root = add_inotify_watch(fd, folder_root);
    if (wd_root == -1) goto end;

    /* remove_watch ./ */
    if(remove_inotify_watch(fd, wd_root)) goto end;
    wd_root = -1;

    /* remove_watch ./scratch */
    if(remove_inotify_watch(fd, wd_scratch)) goto end;
    wd_scratch = -1;

    printf("Run the following command in mdb:\n\n");
    printf("  *inotify_softstate::walk softstate | ::print -t inotify_state_t ins_nevents\n\n");
    printf("When done, press any key to exit ...\n");
    getchar();

end:
    if (wd_root != -1) remove_inotify_watch(fd, wd_root);
    if (wd_scratch != -1) remove_inotify_watch(fd, wd_scratch);
    if (wd_folder1 != -1) remove_inotify_watch(fd, wd_folder1);
    if (wd_folder2 != -1) remove_inotify_watch(fd, wd_folder2);
    if (wd_folder11 != -1) remove_inotify_watch(fd, wd_folder11);
    if (wd_folder21 != -1) remove_inotify_watch(fd, wd_folder21);
    close( fd );

    return (0);
}
```

**Comment by bcantrill** [comment]
Thank you for such a thorough bug report!  It appears to be a logic issue whereby we don't call `pollwakeup` in the explicit removal case -- if the *only* events are `IN_IGNORED`, we won't indicate that we have events to read.  Prototype fix to come...

**Comment by youzhongyang** [comment]
Thanks Bryan.

I tested the following change, it works. Does it look right?
<pre>
diff --git a/usr/src/uts/common/io/inotify.c b/usr/src/uts/common/io/inotify.c
index fa6cf0f..a18f59d 100644
--- a/usr/src/uts/common/io/inotify.c
+++ b/usr/src/uts/common/io/inotify.c
@@ -519,8 +519,10 @@ inotify_watch_event(inotify_watch_t *watch, uint64_t mask, char *name)
        state->ins_nevents++;
        state->ins_size += sizeof (event->ine_event) + len;

-       if (removal)
+       if (removal) {
+               pollwakeup(&state->ins_pollhd, POLLRDNORM | POLLIN);
                return;
+       }

        if ((watch->inw_mask & IN_ONESHOT) && !watch->inw_fired) {
                /*
</pre>

**Comment by bcantrill** [comment]
Simply adding the `pollwakeup` in the removal case is the wrong thing to do because it both causes spurious calls to `pollwakeup`, and it results in calls to `pollwakeup` with driver locks held. The correct fix is similar, though:  we need to add a call to `pollwakeup` after we have dropped the `ins_lock` in `inotify_rm_watch`.  With that fix, the test runs properly.  I'll be fixing this under the aegis of [OS-5714](https://smartos.org/bugview/OS-5714).

**Comment by youzhongyang** [comment]
I see, thank you Bryan.

**Comment by jjelinek** [comment]
Fixed in commit b670ea1
